studio-docs/source/sharing-graphs.mdx

```diff
@@ -0,0 +1,498 @@
+  window.addEventListener("message", onPostMessageReceived);
+}
+```
+
+</MultiCodeBlock>
+
+Now we can import and call `setupEmbedRelay` on any page where we'll embed the Explorer. However, we still need to define the `executeOperation` function that's called by our callback.
+
+#### 3. Define operation execution logic
+
+Our webpages can use the `fetch` API to send operations to our GraphQL server. To do this, we need the URL of our server's GraphQL endpoint. Substitute it in the `executeOperation` function below, where indicated.
+
+<MultiCodeBlock>
+
+```ts:title=setupEmbedRelay.ts
+// Helper function that adds content-type: application/json
+// to each request's headers if not present
+function getHeadersWithContentType(
+  headers: Record<string, string> | undefined
+) {
+  const headersWithContentType = headers ?? {};
+  if (
+    Object.keys(headersWithContentType).every(
+      (key) => key.toLowerCase() !== "content-type"
+    )
+  ) {
+    headersWithContentType["content-type"] = "application/json";
+  }
+  return headersWithContentType;
+}
+
+// Function for executing operations
+async function executeOperation({
+  operation,
+  operationName,
+  variables,
+  headers,
+  embeddedExplorerIFrame,
+  operationId,
+}: {
+  operation: string;
+  operationName?: string;
+  variables?: JSONValue;
+  headers?: Record<string, string>;
+  embeddedExplorerIFrame?: HTMLIFrameElement;
+  operationId: string;
+}) {
+  const response = await fetch(
+    // Substitute your server's URL for this example URL.
+    "https://apollo-fullstack-tutorial.herokuapp.com/",
+    {
+      method: "POST",
+      headers: getHeadersWithContentType(headers),
+      body: JSON.stringify({
+        query: operation,
+        variables,
+        operationName,
+      }),
+    }
+  );
+
+  // After the operation completes, post a response message to the
+  // iframe that includes the response data
+  await response.json().then((response) => {
+    embeddedExplorerIFrame?.contentWindow?.postMessage(
+      {
+        // Include the same operation ID in the response message's name
+        // so the Explorer knows which operation it's associated with
+        name: `${EXPLORER_QUERY_MUTATION_RESPONSE}:${operationId}`,
+        response,
+      },
+      EMBEDDABLE_EXPLORER_URL
+    );
+  });
+}
+```
+
+</MultiCodeBlock>
+
+Instead of hardcoding your server's URL as shown, you can modify `executeOperation` to take a server URL parameter if you'll be embedding Explorers for multiple graphs.
+
+#### 4. Embed the `iframe`
+
+We're ready to add an embedded Explorer to our page! Here's an example `App.tsx` file that uses our `setupEmbedRelay` function to wire up an embedded Explorer `iframe`:
+
+<MultiCodeBlock>
+
+```tsx:title=App.tsx
+import React, { useEffect } from 'react';
+import { EMBEDDABLE_EXPLORER_URL, setUpEmbedRelay } from './setupEmbedRelay';
+
+export const App = ()=> {
+
+  useEffect(() => {
+    setUpEmbedRelay();
```

Comment:
If we are going to include this as a React demo, we might want to do it in a more React-y way (including cleaning up the listener on unmount). I did it this way because I thought people would want to vanilla js in the docs. Thoughts?

Context: https://apollograph.slack.com/archives/C027DK254DS/p1629937720000500?thread_ts=1629935176.086300&cid=C027DK254DS

```
  useEffect(() => {
    const onPostMessageReceived = (event:MessageEvent<{
      name?: string,
      operation?: string,+      operationName?: string,
      variables?: string,
      headers?:string
    }>) => {
      const isQueryOrMutation = event.data.name?.startsWith('ExplorerRequest:');
      const currentOperationId = event.data.name?.split(':')[1]
      if(isQueryOrMutation && event.data.operation) {
          executeOperation({
            operation: event.data.operation,
            operationName: event.data.operationName,
            variables: event.data.variables,
          }).then((response) => {
            const embeddedExplorerIFrame = document.getElementById('embedded-explorer') as HTMLIFrameElement;
            embeddedExplorerIFrame?.contentWindow?.postMessage({
              name: isQueryOrMutation ?
                `ExplorerResponse:${currentOperationId}` :
                `ExplorerSubscriptionResponse:${currentOperationId}`,
              response,
            }, EMBEDDABLE_EXPLORER_URL);
          });
      }
    }
    window.addEventListener('message', onPostMessageReceived);

    return () => window.removeEventListener('message', onPostMessageReceived);
  }, [])

```

I can also make a vanilla js demo it is just more annoying with the babel and package managers etc that create react app does for you